File: 2023/python/day03.py
#!/usr/bin/env python3

import sys
from collections import defaultdict

lines = sys.stdin.readlines()
gear_nums = defaultdict(list)
p1 = p2 = 0

for (y, line) in enumerate(lines):
  num = 0
  sym = False
  gears = set()
  for (x, c) in enumerate(line):
    if c.isdigit():
      num = num*10 + int(c)
      for y2 in [y-1, y, y+1]:
        if y2 < 0: continue
        if y2 >= len(lines): continue
        for x2 in [x-1, x, x+1]:
          if x2 < 0: continue
          if x2 >= len(lines[y2]): continue
          c = lines[y2][x2]
          if not c.isdigit() and c != '.':
            sym = True
            if c == '*':
              gears.add((x2, y2))
    elif num > 0:
      if sym:
        p1 += num
        for gear in gears:
          gear_nums[gear].append(num)
      num = 0
      sym = False
      gears = set()

for nums in gear_nums.values():
  if len(nums) == 2:
    p2 += nums[0] * nums[1]

print("03:", p1, p2)

File: 2023/kotlin/src/Day03.kt
import java.io.File

fun main(args: Array<String>) {
    val lines = File(args.getOrNull(0) ?: "../../inputs/2023/03-input.txt")
        .readLines()

    val gearNums: MutableMap<Int, MutableList<Int>> = mutableMapOf()

    var p1 = 0

    lines.withIndex().forEach { (y, line) ->
        var num = 0
        var sym = false
        val gears: MutableSet<Int> = mutableSetOf()

        // add '.' to end so we don't have to deal with edge case of the last
        // character being a digit
        "$line.".withIndex().forEach { (x, c) ->
            if (c.isDigit()) {
                num = num * 10 + c.toString().toInt()
                sym = sym || (y - 1..y + 1)
                    .filter { it >= 0 && it < lines.count() }
                    .any { y2 ->
                        (x - 1..x + 1)
                            .filter { it >= 0 && it < line.length }
                            .map { lines[y2][it] }
                            .any { it != '.' && !it.isDigit() }
                    }
                gears.addAll((y - 1..y + 1)
                    .filter { it >= 0 && it < lines.count() }
                    .flatMap { y2 -> (x - 1..x + 1)
                        .filter { it >= 0 && it < line.length }
                        .filter { lines[y2][it] == '*' }
                        .map { y2 * 10000 + it } // sue me
                    })
            } else if (num > 0) {
                if (sym) p1 += num
                gears.forEach {
                    val ent = gearNums[it] ?: mutableListOf()
                    ent.add(num)
                    gearNums[it] = ent
                }
                num = 0
                sym = false
                gears.clear()
            }
        }
    }

    val p2: Int = gearNums.values
        .filter { it.count() == 2 }
        .sumOf { it[0] * it[1] }

    println("03: $p1 $p2")
}
